File: src/CopilotDiscordBot/Utils/PathUtils.cs
namespace CopilotDiscordBot.Utils;

public static class PathUtils
{
    public static string NormalizeFullPath(string path)
        => Path.GetFullPath(path.Trim().Trim('"'));

    public static bool IsWithinRoot(string fullPath, string rootFullPath)
    {
        var fp = NormalizeFullPath(fullPath);
        var rp = NormalizeFullPath(rootFullPath);

        if (!rp.EndsWith(Path.DirectorySeparatorChar))
        {
            rp += Path.DirectorySeparatorChar;
        }

        return fp.StartsWith(rp, StringComparison.OrdinalIgnoreCase);
    }

    public static string ResolveRepoPath(string reposRoot, string userInput)
    {
        var input = userInput.Trim().Trim('"');
        if (string.IsNullOrWhiteSpace(input)) throw new ArgumentException("Repo path is required");

        // If user passes a drive-qualified path, accept it (but still enforce root).
        var combined = Path.IsPathRooted(input)
            ? input
            : Path.Combine(reposRoot, input);

        return NormalizeFullPath(combined);
    }
}

File: src/utils/pathUtils.ts
import path from "path";

export function normalizeFullPath(p: string): string {
  return path.resolve(p.trim().replace(/^"|"$/g, ""));
}

export function isWithinRoot(fullPath: string, rootFullPath: string): boolean {
  const fp = normalizeFullPath(fullPath);
  let rp = normalizeFullPath(rootFullPath);

  if (!rp.endsWith(path.sep)) {
    rp += path.sep;
  }

  return fp.toLowerCase().startsWith(rp.toLowerCase());
}

export function resolveRepoPath(reposRoot: string, userInput: string): string {
  const input = userInput.trim().replace(/^"|"$/g, "");
  if (!input) throw new Error("Repo path is required");

  const combined = path.isAbsolute(input)
    ? input
    : path.join(reposRoot, input);

  return normalizeFullPath(combined);
}
